cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,99 @@
+# Kubernetes DNS-Based Service Discovery
+
+## About This Document
+
+This document describes a DNS schema for DNS-based Kubernetes service
+discovery. While service discovery in Kubernetes may be provided via
+other protocols and mechanisms, DNS is very commonly used and is a
+highly recommended add-on. The actual DNS service itself though need not
+be provided by the default Kube-DNS implementation, and so this document
+is intended to provide a baseline for commonality between
+implementations.
+
+## Schema Version
+
+This document describes version 1.0 of the schema.
+
+## Query Responses
+
+Any DNS-based service discovery solution for Kubernetes must support the
+queries described below to be considered compliant with this specification.
+
+In the query names below, values not in angle brackets, `< >`,
+are literals. The meaning of the values in angle brackets are defined
+below or in the description of the specific query response.
+
+`<zone>` = configured cluster domain, e.g. cluster.local
+
+`<ns>` = a Namespace
+
+All comparisons between query names and data in Kubernetes is **case-insensitive**.
+
+In the results column, the following definitions should be used for
+words in _italics_.
+
+- _hostname_
+
+  - For endpoints, in order of precedence:
+    - The value of the endpoint’s `hostname` field.
+    - The value of the `HostName` field stored in the map indexed by the endpoint’s IP address in the `endpoints.beta.kubernetes.io/hostnames-map` annotation.
```

Comment:
I don't think it's really worth documenting this.  The beta needs to go away O(soon)